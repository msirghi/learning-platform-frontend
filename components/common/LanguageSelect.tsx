import React, { useState } from 'react';
import LanguageIcon from '@material-ui/icons/Language';
import IconButton from '@material-ui/core/IconButton';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import Menu from '@material-ui/core/Menu';
import { useTranslation } from '../../i18n';
import { useDispatch } from 'react-redux';
import { setInterfaceLocale } from '../../redux/actions/preference/preferenceAction';

const languageChoice = [
  {
    key: 'en',
    value: 'English'
  },
  {
    key: 'ru',
    value: 'Русский'
  }
];

export const LanguageSelect: React.FC = () => {
  const [notificationMenuOpen, setNotificationMenuOpen] = useState<null | HTMLElement>(null);
  const { i18n } = useTranslation();
  const dispatch = useDispatch();

  const handleChange = (val: string) => {
    dispatch(setInterfaceLocale(val));
    i18n.changeLanguage(val);
    handleClose();
  };

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setNotificationMenuOpen(event.currentTarget);
  };

  const handleClose = () => {
    setNotificationMenuOpen(null);
  };

  return (
    <>
      <IconButton onClick={handleClick}>
        <LanguageIcon color='primary' />
      </IconButton>
      <Menu
        anchorEl={notificationMenuOpen}
        keepMounted
        open={Boolean(notificationMenuOpen)}
        onClose={handleClose}
      >
        <List>
          {languageChoice.map((item) => {
            return (
              <ListItem key={item.key} button onClick={() => handleChange(item.key)}>
                {item.value}
              </ListItem>
            );
          })}
        </List>
      </Menu>
    </>
  );
};
