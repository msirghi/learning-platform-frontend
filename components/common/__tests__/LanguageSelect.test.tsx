import { IconButton, ListItem, Menu } from '@material-ui/core';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { LanguageSelect } from '../LanguageSelect';

describe('LanguageSelect component', () => {
  beforeAll(() => {
    jest.mock('next-i18next', () => ({
      useTranslation: () => {
        return {
          t: (str) => str,
          i18n: {
            changeLanguage: () => new Promise(() => jest.fn())
          }
        };
      }
    }));
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    const wrapper = shallow(<LanguageSelect />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should open menu on language icon click', () => {
    const wrapper = shallow(<LanguageSelect />);
    const button = wrapper.find(IconButton);
    expect(button.length).toBe(1);

    button.simulate('click', { currentTarget: 're' });
    const menu = wrapper.find(Menu);
    expect(menu.props().open).toBeTruthy();
  });

  it('should close menu on language selection', () => {
    const wrapper = shallow(<LanguageSelect />);
    const item = wrapper.find(ListItem).at(0);

    item.simulate('click', { currentTarget: 're' });
    const menu = wrapper.find(Menu);
    expect(menu.props().open).toBeFalsy();
  });
});
