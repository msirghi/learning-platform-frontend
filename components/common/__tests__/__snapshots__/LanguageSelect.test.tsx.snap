// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelect component should match the snapshot 1`] = `
<Fragment>
  <WithStyles(ForwardRef(IconButton))
    onClick={[Function]}
  >
    <Memo([object Object])
      color="primary"
    />
  </WithStyles(ForwardRef(IconButton))>
  <WithStyles(ForwardRef(Menu))
    anchorEl={null}
    keepMounted={true}
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(List))>
      <WithStyles(ForwardRef(ListItem))
        button={true}
        key="en"
        onClick={[Function]}
      >
        English
      </WithStyles(ForwardRef(ListItem))>
      <WithStyles(ForwardRef(ListItem))
        button={true}
        key="ru"
        onClick={[Function]}
      >
        Русский
      </WithStyles(ForwardRef(ListItem))>
    </WithStyles(ForwardRef(List))>
  </WithStyles(ForwardRef(Menu))>
</Fragment>
`;
